[add](管理者側)お問い合わせ一覧閲覧機能、回答機能、対応結果のステータス変更機能の追加

ryuzo111 opened the pull request:
【説明文】
管理者側のお問い合わせ一覧閲覧機能、回答機能、対応結果のステータス変更機能の追加をしました。
以下のURLからログインしてご確認ください。
https://procir-study.site/yajima427/insurance_sns/matching_insurance/public/admin/login
回答してすべての対応が終わったら対応結果ステータスの変更をする想定です。(1回の回答ではご納得頂けない可能性もあるため)

【コマンド】
php artisan migrate:refresh --seed
contactのfactoryを追加したため

【期待される結果】
管理者側で以下の処理をすることが可能になる。
・お問い合わせ一覧を閲覧
・お問い合わせを回答して、回答内容を問い合わせ者にメールで送信
・お問い合わせの対応結果のステータスを変更する事ができる。(未解決→解決)

ご確認お願いいたします！

ryuzo111 committed:
[add]contact機能の追加

footerのボタンから管理者に問い合わせるページを作成。

ryuzo111 committed:
[update]formタグをformファザードに変更

ryuzo111 committed:
[fix]お問い合わせ完了時の遷移先を変更、contactクラスのリファクタリング

お問い合わせ完了時は投稿一覧ページに遷移する仕様に変更

ryuzo111 committed:
[fix]変数名のtypoを修正

ryuzo111 committed:
[fix]ログインユーザーの場合でお問い合わせした際に表示される「メールアドレスは必須です」のエラーを修正

contactモデルの記述を修正

ryuzo111 committed:
[update]adminにsuccessとerrorを追加

ryuzo111 committed:
[add]管理者側で問い合わせ一覧を確認

お問い合わせの内容を表示する事ができる

ryuzo111 committed:
[add]お問い合わせ回答機能とお問い合わせ状況ステータスの変更機能の追加

管理者画面からお問い合わせ対応状況が未解決のものに限り、回答とステータスの変更が可能になる

ryuzo111 committed:
[fix]解決済みの問い合わせには回答と対応ステータスの変更ができなくなる

komo15 commented on resources/views/contact/answer_mail.blade.php:
```diff
@@ -0,0 +1,9 @@
+<h1>お問い合わせについて</h1>
+<p>いつも「ほけんとーく」ご利用いただき誠にありがとうございます。</p>
+<p>以前でお問い合わせ頂いた内容について以下ご回答いたします。</p>
```
「以前に」の方が自然かなと思います。

komo15 commented on resources/views/contact/answer_mail.blade.php:
```diff
@@ -0,0 +1,9 @@
+<h1>お問い合わせについて</h1>
+<p>いつも「ほけんとーく」ご利用いただき誠にありがとうございます。</p>
```
「ほけんとーく」”を”の方が良いかなと思います。

komo15 commented on resources/views/contact/answer_mail.blade.php:
```diff
@@ -0,0 +1,9 @@
+<p>いつも「ほけんとーく」ご利用いただき誠にありがとうございます。</p>
+<p>以前でお問い合わせ頂いた内容について以下ご回答いたします。</p>
+
+{{ $answer }}
```
$answer部分が他の文言と見分けがつきにくいので、行を空けるか、======で囲う等で目立たせた方が良いかなと思いました。

komo15 replied:
すみません！answer_mail.blade.phpでの見え方としては、私も同様なのですが、
実際送られるメールは1行になっているため、回答が長いと見づらさを感じます。
また、回答そのものについて、Formでは改行して記載できますが、
来たメールが1行になってしまっているので、改行が反映できないかなと思いました。

komo15 commented on public/storage:
```diff
@@ -1 +1 @@
-/home/www/hurusawa205/insurance/matching_insurance/storage/app/public
\ No newline at end of file
+/home/www/yajima427/insurance_sns/matching_insurance/storage/app/public
```
これ、私も申し訳なかったのですが、他の人のブランチを取り込むたび変えなければならないのは煩わしいですね・・・私ももう少し確認してみます。

ryuzo111 replied:
あ、ほんとですね、、
毎回毎回だと少し煩わしいですね、、

komo15 commented on resources/views/contact/answer_form.blade.php:
```diff
@@ -0,0 +1,22 @@
+@section('content')
+
+<p>いただいた問い合わせ:{{ $contact->content }}</p>
+
```
使用ユーザーであれば、名前を表示したりプロフィール画面に飛べたりするとより回答しやすいのではと思います。

komo15 commented:
「期待される結果」の部分については、きちんと動作していることを確認しました。
また、大変申し訳ありません！私がプルリクをごちゃごちゃいじったからと思うのですが、
web.phpがコンフリクト起こしています。
おそらく、yajimaさんがインデントを修正してくれているのを私の修正していない状態に
なり替わっているのかと思います。大変お手数おかけし申し訳ないのですが、
こちらyajimaさんに合わせていただけたらと思います。


1点、仕様書に管理者機能については詳細の記載が無いため、仕様の確認になります。

現在、
”解決済でない問い合わせは回答フォームから送れる”
という状態かと思いますが、1回でご納得いただけない場合は
基本的にはメールでのやり取りになるのかなと思います。
この、管理者メールは今後共有メールになるのでしょうか？
現状、フォームから送った履歴がサイト上で確認できないため、
aさんをyajimaさんが対応しているにも関わらず、古澤が気づかずフォームから送ってしまう
ということもあり得るかなと思いました。
また、個人の場合でもうっかりということもあるかと思いますので、
statusで「対応中（基本的にメールでのやり取り）」
等していった方が良いのかなと思いますが、statusがtinyintなので、それは今後実装していく形でしょうか？

rhapsody0919 commented on app/Http/Controllers/ContactController.php:
```diff
@@ -38,4 +40,37 @@ public function contact(SaveContactRule $request)
+        return view('contact.index', compact('contacts'));
+    }
+
+    public function answerForm(Request $request)
```
answerFormだと、「フォームを答える」もしくは、「アンサーフォーム」で、名前が不明確です。

僕のコントローラは、リソースコントローラを作成したので、「edit」とかになっていますが、できたら、showAnswerFormとかがいいと思います。
変数や関数名は本当に大事です。

ryuzo111 committed:
[fix]ご指摘事項を修正

ryuzo111 commented:
修正事項
・メールの文言を変更
・ログイン中のユーザーについてはプロフィールページに遷移することが可能
・回答する際のメッソド名を変更
・statusに「対応中」を追加

komo15 commented:
わざわざ管理者用のプロフィール閲覧ページ作ってくださったんですね💦
プロフィール画面はログイン無しでアクセスできるところなので、
ユーザー用ページで良いかなと思ってました。お手間おかけしすみません！

前回言及漏れておりましたが、現状contact view内がユーザー画面と管理者画面の
両方置いてありますが、区別をつけやすいよう、管理者の方にはadminをつける、
もしくはadminディレクトリに入れた方が良いのではと感じました。
この辺りはterukinaさんも指摘されてなかったので、terukinaさんのご意見も聞きたいです。
一連の「問い合わせ」としていっしょくたで問題無さそうであれば、気にしないでいただいて大丈夫です。

対応状況の追加、ありがとうございます。
対応状況が分かれると、以前のように
”解決済だとFormがそもそも開けないので、開ける/開けないのどちらになる"とならず、
開ける場合として未解決/対応中の2パターンあるので、
回答フォーム画面も対応状況の記載はあった方が良いかなと思います。

ryuzo111 commented:
レビューありがとうございます。
adminのところのファイル名もしくは、ディレクトリに追記する件、terukinaさんからのご意見もお待ちいたします！

ryuzo111 committed:
[fix]回答フォームに回答状況を表示する機能を追加

rhapsody0919 commented:
マージすると、コンフリクトが発生していますね。。。
僕だけですかね？
みなさん、他の人がPRを更新したあと、どうやってレビューされていますか？
そのまま、git fetch origin pull/17/head:feature-contact をすると、エラーが出るんですよね。。だからいつもマージをしているんですけど、


$ git merge  origin/feature-contact
Auto-merging routes/web.php
CONFLICT (content): Merge conflict in routes/web.php
Auto-merging resources/views/contact/index.blade.php
CONFLICT (add/add): Merge conflict in resources/views/contact/index.blade.php
Auto-merging resources/views/contact/answer_mail.blade.php
CONFLICT (add/add): Merge conflict in resources/views/contact/answer_mail.blade.php
Auto-merging resources/views/contact/answer_form.blade.php
CONFLICT (add/add): Merge conflict in resources/views/contact/answer_form.blade.php
Auto-merging public/storage
CONFLICT (content): Merge conflict in public/storage
Auto-merging config/status.php
CONFLICT (add/add): Merge conflict in config/status.php
Auto-merging app/Models/Contact.php
CONFLICT (content): Merge conflict in app/Models/Contact.php
Auto-merging app/Http/Controllers/ContactController.php
CONFLICT (content): Merge conflict in app/Http/Controllers/ContactController.php
Automatic merge failed; fix conflicts and then commit the result.

ryuzo111 commented:
レビューありがとうございます。

私も原因がいまいちわかりません、、
私のpushするときの手順が間違っていたのでしょうか、、？
ちなみに私は以下手順で特定のプルリクを更新しております。

https://qiita.com/hinatades/items/d47dec72a87c5fed50f7

komo15 commented:
回答状況確認しました！ありがとうございます。
私のブランチをマージしたからと思いますが、
User.phpにコンフリクトが発生してしまったので、申し訳ありませんが解消をお願いします。

rhapsody0919 commented on app/Models/Contact.php:
```diff
@@ -24,4 +24,26 @@ public function saveContact($request)
+    public function updateStatusById($id)
+    {
+        $target_contact = $this->findOrFail($id);
+        if ($target_contact->status === 2) {
```
このようなハードコーディングは避けるべきです!
2という数値が何を表しているか書いた本人にしかわからないからです。
あと保守性です!

解決策としては、.envもしくはconfig dirから引っ張ってきた方がいいです!

ryuzo111 replied:
レビューありがとうございます！
勉強になります！

ryuzo111 committed:
[fix]コンフリクト解消、ハードコーティングの解消、ページネーションの追加

User.phpのコンフリクト解消と、ハードコーティングの解消、contact.indexページの一覧表示を10件ずつ表示するためにページネーションを追加

Conflicts:
	app/Models/User.php

komo15 commented:
$unfixed
$in_supported
$resolved
みたいなのを、configのstatus内にそれぞれ定義すれば、そのままconfigから数値を持って来れるので
array_searchを入れる必要がなくなります。色々申し訳ありませんが、こうもできるよって辺り、お伝えさせていただきます！

ryuzo111 committed:
[fix]configの記述方法を変更
